src/build.rs

```diff
@@ -64,15 +66,79 @@ impl BuildSystem {
+            } else if file_type.is_symlink() || file_type.is_file() {
+                fs::remove_file(entry.path()).await.context("Cleaning dist/.current failed")?;
+            }
+        }
+
+        Ok(())
+    }
+
     async fn do_build(&mut self) -> Result<()> {
-        // TODO: delete the contents of the `dist/.current` dir (currently in flight elsewhere).
+        // Ensure the output dist directories are in place.
+        fs::create_dir_all(self.cfg.final_dist.as_path())
+            .await
+            .with_context(|| "error creating build environment directory: dist")?;
+        fs::create_dir_all(self.cfg.staging_dist.as_path())
+            .await
+            .with_context(|| "error creating build environment directory: dist/.current")?;
 
-        // Ensure the output dist directory is in place.
-        fs::create_dir_all(self.cfg.dist.as_path()).await?;
+        self.prepare_staging_dist().await.context("error preparing build environment")?;
 
         // Spawn the source HTML pipeline. This will spawn all other pipelines derived from
         // the source HTML, and will ultimately generate and write the final HTML.
         self.html_pipeline.clone().spawn().await?;
+
+        // Move distrbution from staging dist to final dist
+        self.finalize_dist().await.context("error applying built distribution")?;
+        Ok(())
+    }
+
+    /// Moves the contents of dist/.current into dist, signifying the application
+    /// of a successful build. Also removes dist/.current afterwards.
+    async fn finalize_dist(&self) -> Result<()> {
+        let final_dist = self.cfg.final_dist.clone();
+        let staging_dist = self.cfg.staging_dist.clone();
+        self.progress.clone().set_message("applying new distribution");
+
+        // Build succeeded, so delete everything in `dist`,
+        // copy everything from `dist/.current` to `dist`, and
+        // then delete `dist/.current`.
+        let mut entries = fs::read_dir(&final_dist).await.context("error reading dist dir")?;
+        while let Some(entry) = entries.next().await {
+            let entry = entry.context("error reading contents of dist dir")?;
+            if entry.file_name() == ".current" {
+                continue;
+            }
+
+            let file_type = entry.file_type().await.context("error reading metadata of file in dist dir")?;
+
+            if file_type.is_dir() {
+                remove_dir_all(entry.path().into()).await.context("error cleaning dist")?;
+            } else if file_type.is_symlink() || file_type.is_file() {
+                fs::remove_file(entry.path()).await.context("error cleaning dist")?;
+            }
+        }
+
+        copy_dir_recursive(staging_dist.to_path_buf(), final_dist.to_path_buf())
+            .await
+            .context("error copying dist/.current dir to dist dir")?;
+
+        remove_dir_all(staging_dist).await.context("error deleting dist/.current")?;
+
```

Comment:
Reviewing the changes in this file, a few small things come to mind which should actually prove to simplify things a bit:
- I'm thinking the flow should be:
    - ensure the dist dir exists
    - then ensure that .stage exists & is empty (probably easiest to just kill the dir & recreate)
- for `finalize_dist`, we could just rename the files/directories which will be much faster, as it is typically just an inode operation and no bytes are actually copied/moved.
    - if we do that, then we shouldn't have to delete the staging dir at the end of `finalize_dist`.